### PR TITLE
docs(gitops): adding dependency logic in gitops.md

### PR DESCRIPTION
**Pull request by daniel-hutao**
Signed-off-by: Daniel Hu <tao.hu@merico.dev>

# Summary

Adding dependency logic in gitops.md

# Test

- `config-gitops.yaml`

```sh
tools:
# name of the instance with github-repo-scaffolding-golang
- name: go-webapp-repo
  plugin:
    # kind of the plugin
    kind: github-repo-scaffolding-golang
    # version of the plugin
    # checkout the version from the GitHub releases
    version: 0.2.0
  # options for the plugin
  options:
    # the repo's owner. It should be case-sensitive here; strictly use your GitHub username; please change the value below.
    owner: daniel-hutao
    # the repo which you'd like to create; please change the value below.
    repo: go-webapp
    # the branch of the repo you'd like to hold the code
    branch: main
    # the image repo you'd like to push the container image; please change the value below.
    image_repo: exploitht/go-webapp
- name: jira-github-default
  plugin:
    # kind of this plugin
    kind: jira-github-integ
    # version of the plugin
    # checkout the version from the GitHub releases
    version: 0.2.0
  # optional; if specified, dtm will make sure the dependency is applied first before handling this tool.
  dependsOn: [ "go-webapp-repo.github-repo-scaffolding-golang" ]
  # options for the plugin
  options:
    # the repo's owner
    owner: daniel-hutao
    # the repo where you'd like to setup GitHub Actions
    repo: go-webapp
    # "base url: https://id.atlassian.net"
    jiraBaseUrl: https://JIRA_ID.atlassian.net
    # "need real user email in cloud Jira"
    jiraUserEmail: tao.hu@merico.dev
    # "get it from project url, like 'HEAP' from https://merico.atlassian.net/jira/software/projects/HEAP/pages"
    jiraProjectKey: DT
    # main branch of the repo (to which branch the plugin will submit the workflows)
    branch: main
- name: golang-webapp-actions
  plugin:
    # name of the plugin
    kind: githubactions-golang
    # version of the plugin
    # checkout the version from the GitHub releases
    version: 0.2.0
  # optional; if specified, dtm will make sure the dependency is applied first before handling this tool.
  dependsOn: [ "go-webapp-repo.github-repo-scaffolding-golang" ]
  # options for the plugin
  options:
    # the repo's owner. It should be case-sensitive here; strictly use your GitHub user name; please change the value below.
    owner: daniel-hutao
    # the repo where you'd like to setup GitHub Actions; please change the value below to an existing repo.
    repo: go-webapp
    # programming language specific settings
    language:
      name: go
      # version of the language
      version: "1.17"
    # main branch of the repo (to which branch the plugin will submit the workflows)
    branch: main
    build:
      # default to false
      enable: True
      # build command, OPTIONAL, the given value below is default value
      command: "go build ./..."
    test:
      # default to false
      enable: True
      # test command, OPTIONAL, the given value below is default value
      command: "go test ./..."
      coverage:
        # default to false
        enable: True
        # go test profile subcommand, OPTIONAL, the given value below is default value
        profile: "-race -covermode=atomic"
        output: "coverage.out"
    docker:
      # docker build/push related, default to false
      enable: False
      # dockerhub image repo; please change the value below.
      repo: go-webapp
- name: argocd-dev
  plugin:
    # name of the plugin
    kind: argocd
    # version of the plugin
    # checkout the version from the GitHub releases
    version: 0.2.0
  options:
    # need to create the namespace or not, default: false
    create_namespace: true
    repo:
      # name of the Helm repo
      name: argo
      # url of the Helm repo
      url: https://argoproj.github.io/argo-helm
    # Helm chart information
    chart:
      # name of the chart
      chart_name: argo/argo-cd
      # release name of the chart
      release_name: argocd
      # k8s namespace where ArgoCD will be installed
      namespace: argocd
      # whether to wait for the release to be deployed or not
      wait: true
      # the time to wait for any individual Kubernetes operation (like Jobs for hooks). This defaults to 5m0s
      timeout: 5m
      # whether to perform a CRD upgrade during installation
      upgradeCRDs: true
- name: helloworld
  plugin:
    # name of the plugin
    kind: argocdapp
    # version of the plugin
    # checkout the version from the GitHub releases
    version: 0.2.0
  # if specified, dtm will make sure the dependency is applied first before handling this tool.
  dependsOn: [ "argocd-dev.argocd" ]
  # options for the plugin
  options:
    # information on the ArgoCD application
    app:
      # name of the ArgoCD application
      name: hello
      # where the ArgoCD application CRD will be created
      namespace: argocd
    # destination of the application
    destination:
      # on which server to deploy
      server: https://kubernetes.default.svc
      # in which namespace to deploy
      namespace: default
    # source of the application
    source:
      # which values file to use in the Helm chart
      valuefile: values.yaml
      # path of the Helm chart
      path: charts/go-hello-http
      # Helm chart repo URL, this is only an example, do not use this
      repoURL: https://github.com/ironcore864/openstream-gitops-test.git
```

- `./dtm-darwin-arm64 apply -f config-gitops.yaml`

```sh
2022-03-19 17:06:03 ℹ [INFO]  Apply started.
2022-03-19 17:06:03 ℹ [INFO]  Using dir <.devstream> to store plugins.
2022-03-19 17:06:03 ℹ [INFO]  Tool go-webapp-repo (github-repo-scaffolding-golang) found in config but doesn't exist in the state, will be created.
2022-03-19 17:06:03 ℹ [INFO]  Tool argocd-dev (argocd) found in config but doesn't exist in the state, will be created.
2022-03-19 17:06:03 ℹ [INFO]  Tool jira-github-default (jira-github-integ) found in config but doesn't exist in the state, will be created.
2022-03-19 17:06:03 ℹ [INFO]  Tool golang-webapp-actions (githubactions-golang) found in config but doesn't exist in the state, will be created.
2022-03-19 17:06:03 ℹ [INFO]  Tool helloworld (argocdapp) found in config but doesn't exist in the state, will be created.
Continue? [y/n]
Enter a value (Default is n): y

2022-03-19 17:06:05 ℹ [INFO]  Start executing the plan.
2022-03-19 17:06:05 ℹ [INFO]  Changes count: 5.
2022-03-19 17:06:05 ℹ [INFO]  -------------------- [  Processing progress: 1/5.  ] --------------------
2022-03-19 17:06:05 ℹ [INFO]  Processing: go-webapp-repo(kind: github-repo-scaffolding-golang) -> Create ...
2022-03-19 17:06:08 ℹ [INFO]  The repo go-webapp has been created.
2022-03-19 17:06:21 ✔ [SUCCESS]  Plugin go-webapp-repo(github-repo-scaffolding-golang) Create done.
2022-03-19 17:06:21 ℹ [INFO]  -------------------- [  Processing progress: 2/5.  ] --------------------
2022-03-19 17:06:21 ℹ [INFO]  Processing: argocd-dev(kind: argocd) -> Create ...
2022-03-19 17:06:22 ℹ [INFO]  Creating or updating helm chart ...
2022/03/19 17:06:24 creating 1 resource(s)
2022/03/19 17:06:24 CRD applications.argoproj.io is already present. Skipping.
2022/03/19 17:06:24 creating 1 resource(s)
2022/03/19 17:06:24 CRD applicationsets.argoproj.io is already present. Skipping.
2022/03/19 17:06:24 creating 1 resource(s)
2022/03/19 17:06:24 CRD argocdextensions.argoproj.io is already present. Skipping.
2022/03/19 17:06:24 creating 1 resource(s)
2022/03/19 17:06:24 CRD appprojects.argoproj.io is already present. Skipping.
2022/03/19 17:06:26 creating 44 resource(s)
2022/03/19 17:06:26 beginning wait for 44 resources with timeout of 5m0s
2022/03/19 17:06:26 Deployment is not ready: argocd/argocd-applicationset-controller. 0 out of 1 expected pods are ready
2022/03/19 17:06:28 Deployment is not ready: argocd/argocd-applicationset-controller. 0 out of 1 expected pods are ready
2022/03/19 17:06:30 Deployment is not ready: argocd/argocd-applicationset-controller. 0 out of 1 expected pods are ready
2022/03/19 17:06:32 Deployment is not ready: argocd/argocd-applicationset-controller. 0 out of 1 expected pods are ready
2022/03/19 17:06:34 Deployment is not ready: argocd/argocd-repo-server. 0 out of 1 expected pods are ready
2022/03/19 17:06:36 Deployment is not ready: argocd/argocd-repo-server. 0 out of 1 expected pods are ready
2022/03/19 17:06:38 Deployment is not ready: argocd/argocd-repo-server. 0 out of 1 expected pods are ready
2022/03/19 17:06:40 Deployment is not ready: argocd/argocd-repo-server. 0 out of 1 expected pods are ready
2022/03/19 17:06:42 Deployment is not ready: argocd/argocd-repo-server. 0 out of 1 expected pods are ready
2022/03/19 17:06:44 Deployment is not ready: argocd/argocd-repo-server. 0 out of 1 expected pods are ready
2022/03/19 17:06:47 release installed successfully: argocd/argo-cd-4.2.1
2022-03-19 17:06:47 ✔ [SUCCESS]  Plugin argocd-dev(argocd) Create done.
2022-03-19 17:06:47 ℹ [INFO]  -------------------- [  Processing progress: 3/5.  ] --------------------
2022-03-19 17:06:47 ℹ [INFO]  Processing: jira-github-default(kind: jira-github-integ) -> Create ...
2022-03-19 17:06:48 ℹ [INFO]  Creating GitHub Actions workflow jira-github-integ.yml ...
2022-03-19 17:06:48 ✔ [SUCCESS]  Github Actions workflow jira-github-integ.yml created.
2022-03-19 17:06:50 ✔ [SUCCESS]  Plugin jira-github-default(jira-github-integ) Create done.
2022-03-19 17:06:50 ℹ [INFO]  -------------------- [  Processing progress: 4/5.  ] --------------------
2022-03-19 17:06:50 ℹ [INFO]  Processing: golang-webapp-actions(kind: githubactions-golang) -> Create ...
2022-03-19 17:06:51 ℹ [INFO]  Creating GitHub Actions workflow pr-builder.yml ...
2022-03-19 17:06:51 ✔ [SUCCESS]  Github Actions workflow pr-builder.yml created.
2022-03-19 17:06:52 ℹ [INFO]  Creating GitHub Actions workflow main-builder.yml ...
2022-03-19 17:06:53 ✔ [SUCCESS]  Github Actions workflow main-builder.yml created.
2022-03-19 17:06:53 ✔ [SUCCESS]  Plugin golang-webapp-actions(githubactions-golang) Create done.
2022-03-19 17:06:53 ℹ [INFO]  -------------------- [  Processing progress: 5/5.  ] --------------------
2022-03-19 17:06:53 ℹ [INFO]  Processing: helloworld(kind: argocdapp) -> Create ...
2022-03-19 17:06:53 ℹ [INFO]  application.argoproj.io/hello created
2022-03-19 17:06:53 ✔ [SUCCESS]  Plugin helloworld(argocdapp) Create done.
2022-03-19 17:06:53 ℹ [INFO]  -------------------- [  Processing done.  ] --------------------
2022-03-19 17:06:53 ✔ [SUCCESS]  All plugins applied successfully.
2022-03-19 17:06:53 ✔ [SUCCESS]  Apply finished.
```

### CODE REVIEW

File: docs/plugins/argocdapp_plugin.md [hunk 1]
```diff
@@ -18,8 +20,8 @@ tools:
     # version of the plugin
     # checkout the version from the GitHub releases
     version: 0.2.0
-  # optional; if specified, dtm will make sure the dependency is applied first before handling this tool.
-  dependsOn: [ "TOOL1_NAME.TOOL1_KIND", "TOOL2_NAME.TOOL2_KIND" ]
+  # if specified, dtm will make sure the dependency is applied first before handling this tool.
+  dependsOn: [ "ARGOCD_INSTANCE_NAME.argocd" ]
```
Comment (IronCore864) [inline docs/plugins/argocdapp_plugin.md:24]: INSTANCE -> TOOL? Because we call it TOOL_NAME across the doc/code.

Comment (daniel-hutao) [reply docs/plugins/argocdapp_plugin.md:24]: I think the ARGOCD_INSTANCE_NAME == TOOL_NAME, but ARGOCD_TOOL_NAME is not very good.
But TOOL_NAME is a little unclear.

File: docs/plugins/argocdapp_plugin.md [hunk 2]
```diff
@@ -3,10 +3,12 @@
 This plugin creates an [ArgoCD Application](https://argo-cd.readthedocs.io/en/stable/core_concepts/) custom resource.
 
 **Notes:**
-- ArgoCD itself must have been already installed before the usage of this plugin. To install ArgoCD, use the [argocd plugin](https://github.com/merico-dev/stream/blob/main/docs/argocd_plugin.md).
+- ArgoCD itself must have been already installed before the usage of this plugin.
+  To install ArgoCD, use the [argocd plugin](https://github.com/merico-dev/stream/blob/main/docs/argocd_plugin.md).
```
Comment (lfbdev) [inline docs/plugins/argocdapp_plugin.md:7]: access "https://github.com/merico-dev/stream/blob/main/docs/argocd_plugin.md" 404 failed

File: docs/plugins/argocdapp_plugin.md (hunk 1 above)
Comment (IronCore864) [reply docs/plugins/argocdapp_plugin.md:24]: "Instance" means something specific. For example, in a highly-available setup, an "instance" might mean one of the ArgoCD server pods. So, "ARGOCD_INSTANCE_NAME" has ambiguity. Please think of a better name. If not possible, add some description about it, either in the example YAML as a comment, or as an extra sentence in the doc.

File: docs/plugins/argocdapp_plugin.md (hunk 2 above)
Comment (lfbdev) [reply docs/plugins/argocdapp_plugin.md:7]: may use  https://github.com/merico-dev/stream/blob/main/docs/plugins/argocd_plugin.md  instead

**Review by IronCore864 — APPROVED** [review]

**Pull request merged**